File: day04/day4.rb
# frozen_string_literal: true

# day4
class PasswordProcessor
  def initialize(file, validator)
    @data = File.read(file).split("\n\n").map do |passport|
      passport.gsub("\n", ' ').split.each_with_object({}) do |field, obj|
        key, value = field.split(':')
        obj[key.to_sym] = value
        obj
      end
    end
    @validator = validator
  end

  VALID_FIELDS = %i[byr iyr eyr hgt hcl ecl pid].freeze

  def count_valid_fields
    @data.count { |passport| VALID_FIELDS.difference(passport.keys).none? }
  end

  def count_valid_passports
    @data.count do |passport|
      VALID_FIELDS.difference(passport.keys).none? &&
        passport.all? { |(field, info)| @validator.field_is_valid?(field, info) }
    end
  end
end

class Validator
  def initialize(field, info)
    @field = field
    @info = info
  end

  def self.field_is_valid?(field, info)
    new(field, info).send(field)
  end

  private

  attr_reader :field, :info

  def byr
    info.between?('1920', '2002')
  end

  def iyr
    info.between?('2010', '2020')
  end

  def eyr
    info.between?('2020', '2030')
  end

  def hgt
    unit = info[-2..].to_sym
    {
      cm: info[0..2].between?('150', '193'),
      in: info[0..1].between?('59', '76')
    }[unit]
  end

  def hcl
    info.match?(/^#[[:xdigit:]]{6}$/)
  end

  def ecl
    %w[amb blu brn gry grn hzl oth].include?(info)
  end

  def pid
    info.match?(/^[[:digit:]]{9}$/)
  end

  def cid
    true
  end
end

processor = PasswordProcessor.new(*$*, Validator)

p processor.count_valid_fields # -> 247
p processor.count_valid_passports # -> 145

File: day04/day4.js
const fs = require('fs');
const readFile = (path) => fs.readFileSync(path, { encoding: 'utf8' });

const PassportProcessor = (file) => {
  const data = readFile(file).split('\n\n').map((passport) => {
    return passport.replace(/\n/g, ' ').split(' ').reduce((map, field) => {
      const [key, value] = field.split(':');
      map.set(key, value);
      return map;
    }, new Map());
  });

  const validFields = ['byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid'];
  const eyeColors = ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth'];

  const hasAllFields = (passport) =>
    validFields.every((field) => passport.has(field));

  const hasValidValues = (passport) => {
    const validators = {
      byr: (data) => 1920 <= data && data <= 2002,
      iyr: (data) => 2010 <= data && data <= 2020,
      eyr: (data) => 2020 <= data && data <= 2030,
      hgt: (data) => {
        if (data.slice(-2) === 'cm') {
          const number = data.slice(0, 3);
          return data.length === 5 && 150 <= number && number <= 193;
        }
        if (data.slice(-2) === 'in') {
          const number = data.slice(0, 2);
          return data.length === 4 && 59 <= number && number <= 76;
        }
        return false;
      },
      hcl: (data) => data.match(/^#[a-f0-9]{6}$/),
      ecl: (data) => eyeColors.includes(data),
      pid: (data) => data.match(/^\d{9}$/),
      cid: (_data) => true,
    }

    return validFields.every((field) => {
      const value = passport.get(field);
      return validators[field](value);
    })
  }

  const countValidPassports = () => {
    return data.reduce((validCount, passport) => {
      if (hasAllFields(passport) && hasValidValues(passport)) {
        validCount++;
      }
      return validCount;
    }, 0);
  }

  const countValidFields = () => {
    return data.reduce((validCount, passport) =>
      hasAllFields(passport) ? validCount + 1 : validCount, 0);
  }

  return {
    countValidFields,
    countValidPassports
  };
}

const processor = PassportProcessor(process.argv[2]);

processor.countValidFields(); // 247
processor.countValidPassports(); // 145
